magpie/api/management/register/register_utils.py

```diff
@@ -79,3 +139,219 @@ def get_discoverable_group_by_name(group_name, db_session):
+
+    .. seealso::
+        - See :ref:`proc_user_registration` for the procedure step details.
+        - :func:`complete_user_registration` for step 5 following approval.
+    """
+    if tmp_token.operation == TokenOperation.USER_REGISTRATION_ADMIN_APPROVE:
+        msg = "Pending user registration was successfully approved."
+        complete_user_registration(tmp_token, request)
+    elif tmp_token.operation == TokenOperation.USER_REGISTRATION_ADMIN_DECLINE:
+        # flush the pending user, this should cascade remove any associated temporary tokens
+        ax.evaluate_call(lambda: request.db.delete(tmp_token.user), fallback=lambda: request.db.rollback(),
+                         content={"user": uf.format_user(tmp_token.user)},
+                         http_error=HTTPInternalServerError, msg_on_fail="Failed deletion of pending user.")
+        msg = "Pending user registration was successfully declined. Pending user has been deleted."
+    else:
+        msg = "Unknown operation received during pending user registration decision by administrator."
+        ax.raise_http(HTTPInternalServerError, detail=msg, content=tmp_token.json())
+    return BaseViews(request).render("magpie.ui.home:templates/message.mako", {"message": msg})
+
+
+def complete_user_registration(tmp_token, request):
+    # type: (TemporaryToken, Request) -> None
+    """
+    Completes the successful user registration following any required validation steps.
+
+    Generates the :term:`User` from the :term:`Pending User`.
+    Then, sends configured notification emails about successful :term:`User` creation.
+
+    Implements steps (5) and (6) of the user registration procedure.
+
+    .. seealso::
+        - See :ref:`proc_user_registration` for the procedure step details.
+        - :func:`register_pending_user` for initial steps that started the process.
+        - :func:`request_admin_approval` for intermediate steps if approval feature was enabled.
+    """
+    LOGGER.debug("[User Registration - Step 5] pending user: [%s]", tmp_token.user.user_name)
+
+    # detach pending user from temporary token to avoid db integrity error since it will become invalid after upgrade
+    pending_user = tmp_token.user
+    tmp_token.user_pending_id = None
+    user = pending_user.upgrade(db_session=request.db)
+    LOGGER.debug("Pending user upgraded to full user: [%s (%s)]", user.user_name, user.id)
+
+    # notify the user of its successful account validation, approval and creation
+    params = {"user": user}
+    template = get_email_template("MAGPIE_USER_REGISTRATION_COMPLETED_EMAIL_TEMPLATE", request)
+    ax.evaluate_call(lambda: send_email(user.email, request, template, params),
+                     fallback=lambda: request.db.rollback(), http_error=HTTPInternalServerError,
+                     msg_on_fail="Error occurred during user registration when trying to send "
+                                 "email to pending user for confirmation of its submitted email.")
```

Comment:
I'm using this wording because it refers to the email confirmation step of the full *user registration procedure*, which itself involves many back-and-forth emails. Wording "confirmation of registration" could be confused with confirmation/approval by the admin of that registration.